python/CMakeLists.txt

```diff
@@ -855,6 +856,10 @@ set(CYTHON_FLAGS "${CYTHON_FLAGS}" "--warning-errors")
 # undocumented Cython feature.
 set(CYTHON_FLAGS "${CYTHON_FLAGS}" "--no-c-in-traceback")
 
+if(CYTHON_VERSION VERSION_GREATER_EQUAL "3.1.0a0")
+  set(CYTHON_FLAGS "${CYTHON_FLAGS}" "-Xfreethreading_compatible=True")
```

Comment:
Is the flag documented somewhere?

Reply:
It's documented here: https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#compiler-directives.

Reply:
Ok, thanks. Is it useful to enable it in PyArrow? Most performance-critical code in PyArrow is inside Arrow C++, which we usually call with the GIL released anyway.

Reply:
Whether PyArrow releases the GIL or not does not come into play here.

Declaring the `freethreading_compatible` directive prevents the GIL from being re-enbaled by the runtime. If it were to not be added, then any downstream code doing `import pyarrow` would re-enable the GIL, because PyArrow has not declared that it supports free-threading.

Reply:
You mean it would re-enable the GIL for the whole process lifetime? Or only when calling into PyArrow?

Reply:
For the whole process lifetime.

Reply:
Ah, I understand. Thank you!